orgs_settings: Add modal for message content edit time limits.

shubhamdhama opened the pull request:
This adds a modal to edit the time limit to edit message content under "Organisation Settings" section.
Fixes: #6809.
Though the scope of the issue was to add modal for all time limits in settings which I figured out that there are 2: one under "Organisation Settings->Message editing" and another under "Organisation Permission->Streams & custom emoji" but currently, this adds the modal for Message edit limit only, I want the feedback regarding this and also guidance regarding the tests.
I have manually tested this on Chrome and Firefox (in Linux) and this works as expected.
@timabbott @rishig .
![peek 2017-10-29 19-10](https://user-images.githubusercontent.com/22238472/32144311-f73ba36a-bcdc-11e7-95ff-365e6217d550.gif)

rishig commented:
This is looking great, thanks @shubhamdhama!

I think my description may have been confusing; it would be great to have both a text box for entering a number, and a dropdown by which the user can choose minutes, hours, or days.

Any ideas for the design of this, @brockwhittaker? In particular, how to incorporate the "No limit" part.

rishig commented:
By default let's just do a radio button. 

```
[ ] No limit
[ ] Limit to [text input] [dropdown]
```
where text input has space for a 3 digit integer, and dropdown has [minutes, hours, days].
For simplicity, if someone submits with 0 in the text box it should just change it to 1. 
(I would also get rid of the up and down arrows, just a text box is fine).

shubhamdhama commented:
Thanks for the review, I'll be updating PR soon.

rishig commented:
I thought of a better solution, inspired by 
![image](https://user-images.githubusercontent.com/890911/33001286-ac444476-cd61-11e7-8775-6b2dba259034.png)
further down that settings page (it might look slightly different in master, but the basic idea is there in master as well).

So how about:
Message edit limit: `60 minutes :pencil2:` [No limit] 

If you click on the 60 minutes part, it should open a popover (similar in styling to the one for the New stream notifications), where you can enter in a number, and select units from a dropdown.
The No limit link should work similarly to the Disable link for New stream notifications, except without the color change. 

It can have a save icon at the end: :floppy_disk: (whatever the equivalent is in font awesome), and clicking outside the popover should also save it, I think. 

Remember to test the case where it says `No limit :pencil2:`, and you click on it (it should just prefill with the default 60 minutes (or whatever the default is).

Thanks!

shubhamdhama commented:
Updated the PR and looking like working approximately as required above. Tests are failing(which I'll update soon.) Review ( mostly regarding the refactoring of code as I'm not fully confident about the variable names, line comments etc).
@rishig can please review this again.

rishig commented on static/js/settings_org.js:
```diff
@@ -134,6 +137,46 @@ exports.render_notifications_stream_ui = function (stream_id) {
     elem.removeClass('text-warning');
 };
 
+function conversion_form_seconds(time_in_seconds) {
```
We can call this `convert_to_settings_format(seconds)`

rishig commented on static/js/settings_org.js:
```diff
@@ -134,6 +137,46 @@ exports.render_notifications_stream_ui = function (stream_id) {
 };
 
+function conversion_form_seconds(time_in_seconds) {
+    var unit_time_object = {
```
Maybe `settings_time_object`?

rishig commented on static/js/settings_org.js:
```diff
@@ -134,6 +137,46 @@ exports.render_notifications_stream_ui = function (stream_id) {
+    }
+    return unit_time_object;
+}
+exports.update_message_edit_time_limit_settings = function (unit_time_object) {
```
unit_time_object -> settings_time_object

rishig commented on static/js/settings_org.js:
```diff
@@ -6,6 +6,9 @@ var meta = {
 };
 
+// Realm.DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS
+var DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_MINUTES = 10;
```
I think we should just make this DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS

rishig commented on static/js/settings_org.js:
```diff
@@ -134,6 +137,46 @@ exports.render_notifications_stream_ui = function (stream_id) {
+    return unit_time_object;
+}
+exports.update_message_edit_time_limit_settings = function (unit_time_object) {
+    // Updates the message_edit_limit_modal with appropriate values stored in unit_time_object
```
I would do something like
`if settings_time_object is null, settings_time_object = convert_to_settings_format(DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS)`

That will also be clear enough that we won't need a comment.

rishig commented on static/js/settings_org.js:
```diff
@@ -404,28 +483,39 @@ function _set_up() {
         // If allow_message_editing is unchecked, message_content_edit_limit_minutes
         // is irrelevant.  Hence if allow_message_editing is unchecked, and
         // message_content_edit_limit_minutes is poorly formed, we set the latter to
         // a default value to prevent the server from returning an error.
```
is this comment or the code under it still relevant? There should just be no way to save it with an invalid value

rishig commented on static/js/settings_org.js:
```diff
@@ -437,23 +527,27 @@ function _set_up() {
-                                i18n.t("Users can now edit topics for all their messages, and the content of messages which are less than __num_minutes__ minutes old.",
-                                       {num_minutes : data_message_content_edit_limit_minutes}),
-                                message_editing_status);
+                                i18n.t("Users can now edit topics for all their messages, and the content of messages which are less than __num_time__ __unit__ old.",
```
This is going to be complicated to translate. With a modal I think we need no success message at all, since it's visibly clear that the setting was changed, so let's just remove this.

shubhamdhama replied:
So this means `admin-realm-message-editing-status` is of no use further as all changes are visible on button.

rishig commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -26,16 +26,23 @@
-                    {{t 'Message edit limit in minutes (0 for no limit)' }}
+                <label for="change_content_edit_limit"
+                    id="change_content_edit_button_label"
+                    title="{{t 'If non-zero, users can edit their message for this many minutes after it is sent. If No Limit (i.e. zero), users can edit all their past messages.' }}">
```
This is now considerably clearer, so we can just have the title text be 'Message edit limit" to reduce translation burden.

rishig commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -26,16 +26,23 @@
+                <label for="change_content_edit_limit"
+                    id="change_content_edit_button_label"
+                    title="{{t 'If non-zero, users can edit their message for this many minutes after it is sent. If No Limit (i.e. zero), users can edit all their past messages.' }}">
+                    {{t 'Message edit limit :' }}
```
No space before the `:`

shubhamdhama commented:
@rishig while fixing above issues, I found a bug that if there is `No limit` set and if we click `Save Changes` button then it changes the limit to default as in the background we have filled the modal with default values.
Solution to this would be to leave the time value in the modal set to zero(i.e. not updating modal to default values) or to somehow prevent saving the messaging edit limit settings from `Save Changes` button.

rishig commented:
Good catch. Another possible hack would be 
* Clicking "No limit" sets it to 0
* When you open the modal, it checks if the time limit is 0. If it is, then it replaces it wit the default values.

Would that work?

rishig commented:
[This is on hold for three weeks due to exams.]

shubhamdhama commented:
@rishig I've updated the PR, it is working as desired in my system. Can you this review again. And I've updated the node tests to cover maximum changes(though my all changes aren't covered yet) and node tests are passing. [Some unrelated error is failing the build]

rishig commented:
Thanks @shubhamdhama! A few things:
* If someone gives a value that is 0, negative, or malformed, the error message can just be "Failed: Must be a positive number"

hm .. I had few more quick comments, but rebasing on master seems to have broken this for me? 
![image](https://user-images.githubusercontent.com/890911/34189932-e38b2c12-e4f2-11e7-80d3-733ca688a4ed.png)

Any idea what could have happened?

shubhamdhama commented:
@rishig Updated the PR, and it is working fine even after rebasing(might need clearing cache).

rishig commented:
Thanks! Two comments:
* I feel like saving with "0 hours" still turns it into "No limit", rather than leaving what was there originally.
* I don't see the error (or success) messages?

shubhamdhama commented:
> I don't see the error (or success) messages?

I may have misinterpreted this review comment https://github.com/zulip/zulip/pull/7214#discussion_r153059814
So as of now, it shows error success messages on:
* changing the limit if it is successful, it is reflected on the button
* If we disable or enable "message editing" then it shows the success banner.
* If we set a negative or malformed value as edit limit.

It seems like having success/error messages is the most intuitive way, so I'll update it, Can you please confirm whether we want to have them or not?
>I feel like saving with "0 hours" still turns it into "No limit", rather than leaving what was there originally?

Sorry, I forgot to discuss it, but I thought to keep the legacy behavior of this. But, if we want [No limit] as the only way to set No limit, then I will update it as well.

rishig commented:
> misinterpreted review comment

hm .. I see. You did interpret it correctly, but now that I play with it I think I was wrong. Is it easy to put back? A simple message is fine, like "Setting updated", since we're getting rid of all these banners very soon. 

> 0 hours

yeah, this should either be an error, or uncheck the setting, whichever is easier.

shubhamdhama commented:
I've made both the changes as asked, and most of my changes are covered by corresponding node tests with some other corner cases(They all are passing),
![updated_modal](https://user-images.githubusercontent.com/22238472/34452659-9d22fe1e-ed6a-11e7-9b57-c786db98dbb3.gif)

but in my last push, Casper tests were failing, they should not be affected by my changes as I haven't made any change to `organization default language settings` and even to the worst, I'm facing some strange errors while reproducing them.
[Here](https://chat.zulip.org/#narrow/stream/development.20help/subject/Casper.20Test.20Failing.20.237214/near/415679) I've asked help about it

shubhamdhama commented:
Resolved the merge conflicts, but still, I've the same problems of casper test, as mentioned above.

shubhamdhama commented:
Resolved the conflicts and fixed all the tests(Finally!!).
@rishig FYI

timabbott commented:
Closing, since I think we're hoping to change this to a dropdown instead.  You can chat with @rishig about the next steps for this when we discuss the settings roadmap.  One small note is that this is a blocker for working on https://github.com/zulip/zulip/issues/7344, so we should do that after this.

rishig commented:
Thanks for all the work here @shubhamdhama! After playing with it, I feel like it has a lot of visual design complication supporting a corner use case, and I think the dropdown approach will do a better job of hiding the complexity. We can chat about the settings roadmap when we're next both online.

shubhamdhama commented:
Okay, actually I agree this PR become irrelevant now(BTW I've learned a lot about settings subsystem from this PR). Thanks, Rishi.
I'll contact whenever both online.
